libraries/GCS_MAVLink/GCS_MAVFTP.cpp

```diff
@@ -0,0 +1,713 @@
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/Device.h>
+#include <AP_HAL/I2CDevice.h>
+#include "GCS.h"
+#include <AP_Math/crc.h>
+
+#if HAL_OS_POSIX_IO || HAL_OS_FATFS_IO
+
+#if HAL_OS_POSIX_IO
+#include <unistd.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <assert.h>
+#include <stdio.h>
+#include <time.h>
+#include <dirent.h>
+#if defined(__APPLE__) && defined(__MACH__)
+#include <sys/param.h>
+#include <sys/mount.h>
+#else
+#include <sys/statfs.h>
+#endif
+#endif
+
+#if HAL_OS_FATFS_IO
+#include <stdio.h>
+#endif
+
+#define MAX_FTP_DATA_LENGTH (MAVLINK_MSG_FILE_TRANSFER_PROTOCOL_FIELD_PAYLOAD_LEN - sizeof(FTPPayloadHeader))
+
+extern const AP_HAL::HAL& hal;
+
+static const char	kDirentFile = 'F';	///< Identifies File returned from List command
+static const char	kDirentDir = 'D';	///< Identifies Directory returned from List command
+static const char	kDirentSkip = 'S';	///< Identifies Skipped entry from List command
+
+
+void GCS_MAVLINK::_ftp_init()
+{
+    strcpy(_ftp_current_abs_path, HAL_BOARD_STORAGE_DIRECTORY);
+    _session_info.fd = -1;
+    _session_info.stream_download = false;
+}
+
+char *GCS_MAVLINK::_data_as_cstring(FTPPayloadHeader *payload)
+{
+    // guarantee nul termination
+    if (payload->size < MAX_FTP_DATA_LENGTH) {
+        ((char*)payload->data)[payload->size] = '\0';
+
+    }
+    else {
+        payload->data[MAX_FTP_DATA_LENGTH - 1] = '\0';
+    }
+
+    // and return data
+    return (char *) & (payload->data[0]);
+}
+
+void GCS_MAVLINK::_set_current_abs_path(char* rel_path)
+{
+    if (rel_path[0] == '/' && rel_path[1] == '\0') {
+        _ftp_current_abs_path[strlen(HAL_BOARD_STORAGE_DIRECTORY)] = '\0';
+    }
+    else {
+        strcpy(&_ftp_current_abs_path[strlen(HAL_BOARD_STORAGE_DIRECTORY)], rel_path);
+    }
+}
+
+void GCS_MAVLINK::handle_ftp_messages(mavlink_message_t *msg)
+{
+    mavlink_file_transfer_protocol_t ftp_request;
+    mavlink_msg_file_transfer_protocol_decode(msg, &ftp_request);
+    FTPPayloadHeader *payload = reinterpret_cast<FTPPayloadHeader *>(&ftp_request.payload[0]);
+    FTPPayloadHeader *response_payload = reinterpret_cast<FTPPayloadHeader *>(&_ftp_response.payload[0]);
+    uint8_t target_system_id = msg->sysid;
+    FTPErrorCode errorCode = kErrNone;
+    _ftp_errno = 0;
+    // basic sanity checks; must validate length before use
+    if (payload->size > MAX_FTP_DATA_LENGTH) {
+        errorCode = kErrInvalidDataSize;
+        goto out;
+    }
+
+    // check the sequence number: if this is a resent request, resend the last response
+    if (payload->seq_number + 1 == response_payload->seq_number) {
+        // this is the same request as the one we replied to last. It means the (n)ack got lost, and the GCS
+        // resent the request
+        mavlink_msg_file_transfer_protocol_send_struct(chan, &_ftp_response);
+        return;
+    }
+    switch (payload->opcode) {
```

Comment:
a memset of the responce_payload.data would be nice, as mavlink2 packet trimming will work, and save alot of bandwidth

Reply:
definately